namer/namer.cc

```diff
@@ -591,181 +923,698 @@ class NameInserter {
                         sym.show(ctx), symArg.name.show(ctx), methodArg.local._name.show(ctx));
                     e.addErrorLine(sym.data(ctx)->loc(), "Previous definition");
                 }
-                return;
+                return;
+            }
+        }
+    }
+
+    core::SymbolRef defineMethod(core::MutableContext ctx, const FoundMethod &method) {
+        core::SymbolRef owner = methodOwner(ctx, method.flags);
+
+        // There are three symbols in play here, because there's:
+        //
+        // - sym, the symbol which, if it exists, is the 'correct' symbol for the method in question, as it matches
+        //   in both name and argument structure
+        // - currentSym, which is the whatever the non-mangled name currently points to in that context
+        // - replacedSym, which is whatever name was the name that sym replaced.
+        //
+        // In the context of something like
+        //   def f(); end
+        //   def f(x); end
+        //   def f(x, y); end
+        // when the namer is in incremental mode and looking at `def f(x)`, then `sym` refers to `f(x)`,
+        // `currentSym` refers to `def f(x, y)` (which is the symbol that is "currently" non-mangled and available
+        // under the name `f`, because the namer has already run in this context) and `replacedSym` refers to `def
+        // f()`, because that's the symbol that `f` referred to immediately before `def f(x)` became the
+        // then-current symbol. If we weren't running in incremental mode, then `sym` wouldn't exist yet (we would
+        // be about to create it!) and `currentSym` would be `def f()`, because we have not yet progressed far
+        // enough in the file to see any other definition of `f`.
+        auto &parsedArgs = method.parsedArgs;
+        auto sym = ctx.state.lookupMethodSymbolWithHash(owner, method.name, ast::ArgParsing::hashArgs(ctx, parsedArgs));
+        auto currentSym = ctx.state.lookupMethodSymbol(owner, method.name);
```

Comment:
Can these two lookups be combined?
Is it true that if `!currentSym.exists()` than we know for sure `!sym.exists()`? Afaik this is also the common case and thus worth optimizing for

Reply:
Also, in the same vein of optimizing for common case, can this do `enterMethodSymbol` instead to make sure it _now_ exists? and then we can figure out if it existed before based on id

Reply:
This would reduce number of lookups in common case by ~3x